src/sentry/sentry_metrics/indexer/limiters/cardinality.py

```diff
@@ -0,0 +1,128 @@
+from __future__ import annotations
+
+import dataclasses
+from collections import defaultdict
+from typing import Any, Generic, Mapping, MutableMapping, Optional, Sequence, TypedDict, TypeVar
+
+from sentry import options
+from sentry.ratelimits.cardinality import (
+    CardinalityLimiter,
+    GrantedQuota,
+    Quota,
+    RedisCardinalityLimiter,
+    RequestedQuota,
+    Timestamp,
+)
+from sentry.sentry_metrics.configuration import UseCaseKey, get_ingest_config
+from sentry.utils import metrics
+from sentry.utils.hashlib import hash_values
+
+OrgId = int
+
+TMessageKey = TypeVar("TMessageKey")
+
+
+@dataclasses.dataclass(frozen=True)
+class CardinalityLimiterState(Generic[TMessageKey]):
+    _cardinality_limiter: CardinalityLimiter
+    _use_case_id: UseCaseKey
+    _requests: Sequence[RequestedQuota]
+    _grants: Sequence[GrantedQuota]
+    _timestamp: Timestamp
+    keys_to_remove: Sequence[TMessageKey]
+
+
+def _build_quota_key(use_case_id: UseCaseKey, org_id: Optional[OrgId]) -> str:
+    if org_id is not None:
+        return f"metrics-indexer-cardinality-{use_case_id}-org-{org_id}"
+    else:
+        return f"metrics-indexer-cardinality-{use_case_id}-global"
+
+
+@metrics.wraps("sentry_metrics.indexer.construct_quotas")
+def _construct_quotas(use_case_id: UseCaseKey) -> Sequence[Quota]:
+    """
+    Construct write limit's quotas based on current sentry options.
+
+    This value can potentially cached globally as long as it is invalidated
+    when sentry.options are.
+    """
+    if use_case_id == UseCaseKey.PERFORMANCE:
+        return [
+            Quota(**args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.performance.per-org")
+        ]
+    elif use_case_id == UseCaseKey.RELEASE_HEALTH:
+        return [
+            Quota(**args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.releasehealth.per-org")
+        ]
+    else:
+        raise ValueError(use_case_id)
+
+
+class InboundMessage(TypedDict):
+    # Note: This is only the subset of fields we access in this file.
+    org_id: int
+    name: str
+    type: str
+    tags: Mapping[str, str]
+
+
+class TimeseriesCardinalityLimiter:
+    def __init__(self) -> None:
+        self.rate_limiters: MutableMapping[UseCaseKey, CardinalityLimiter] = {}
+
+    def _get_rate_limiter(self, use_case_id: UseCaseKey) -> CardinalityLimiter:
+        if use_case_id not in self.rate_limiters:
+            options = get_ingest_config(use_case_id).cardinality_limiter_cluster_options
+            self.rate_limiters[use_case_id] = RedisCardinalityLimiter(**options)
+
+        return self.rate_limiters[use_case_id]
```

Comment:
Please watch out for the changes @MeredithAnya  did on how to instantiate a rate limtier. That is in master now.

Reply:
ah yes, it's great that this got cleaned up 👍